fs/nova/balloc.c

```diff
@@ -601,65 +605,142 @@ static long nova_alloc_blocks_in_free_list(struct super_block *sb,
 	else
 		temp = &(free_list->last_node->node);
 
-	while (temp) {
-		step++;
-		curr = container_of(temp, struct nova_range_node, node);
+	/* Try huge block allocation for data blocks first */
+	if (IS_DATABLOCKS_2MB_ALIGNED(num_blocks, atype)) {
+		unsigned int left_margin;
+		unsigned int right_margin;
+
+		while (temp) {
+			step++;
+			curr = container_of(temp, struct nova_range_node, node);
+
+			if (!nova_range_node_checksum_ok(curr)) {
+				nova_err(sb, "%s curr failed\n", __func__);
+				goto next;
+			}
 
-		if (!nova_range_node_checksum_ok(curr)) {
-			nova_err(sb, "%s curr failed\n", __func__);
-			goto next;
+			curr_blocks = curr->range_high - curr->range_low + 1;
+			left_margin = 512 - (curr->range_low & 511);
+			right_margin = curr_blocks - left_margin - num_blocks;
+
+			/* do the allocaiton once remaining block is large engough*/
+			if (num_blocks <= (curr_blocks - left_margin)) {
+				struct nova_range_node *node;
+
+				nova_dbgv("curr:%p: num_blocks:%lu curr->range_low:%lu high:%lu", curr, num_blocks, curr->range_low, curr->range_high);
+				/* Free curr before insert the left,right margin node. */
+				rb_erase(&curr->node, tree);
+				free_list->num_blocknode--;
+				if (left_margin) {
+					node = nova_alloc_blocknode(sb);
```

Comment:
I am afraid you cannot call this here because we are holding a spinlock and this method can sleep.

Reply:
That should work.